### PR TITLE
parametrize offsets tests

### PR DESCRIPTION
**Pull request by jbrockmendel**
This began as part of #18489 and quickly became huge and tedious.  Merits separating out and getting out of the way.

- Move Year, Quarter, and Month offset test cases to a dedicated file

- Implement pytest.mark.parametrize in very-nearly all the remaining places.

### CODE REVIEW

**Comment by jreback** [comment]
might be worthwhile moving Week, LastWeekOfMonth, SemiMonth to separate test file
and FY253 to a separate file as well.

**Comment by jbrockmendel** [comment]
> might be worthwhile moving Week, LastWeekOfMonth, SemiMonth to separate test file
and FY253 to a separate file as well.

Very much on board.  I'll also 

1) take the opportunity to move the SemiMonth offsets below the CustomBusinessMonth offsets so they're not in between BusinessMonth and CustomBusinessMonth.

2) throw in some new asvs

File: pandas/tseries/offsets.py [hunk 1]
```diff
@@ -985,6 +985,158 @@ class BusinessMonthBegin(MonthOffset):
...
+    def m_offset(self):
+        kwds = self.kwds
+        kwds = {key: kwds[key] for key in kwds
+                if key not in ['calendar', 'weekmask', 'holidays', 'offset']}
+        return MonthEnd(n=1, normalize=self.normalize, **kwds)
+
+    @apply_wraps
+    def apply(self, other):
+        n = self.n
+        # First move to month offset
```
Comment (jreback) [inline pandas/tseries/offsets.py]: blank lines before comments

File: pandas/tseries/offsets.py [hunk 2]
```diff
@@ -985,6 +985,158 @@ class BusinessMonthBegin(MonthOffset):
...
+        kwds = self.kwds
+        kwds = {key: kwds[key] for key in kwds
+                if key not in ['calendar', 'weekmask', 'holidays', 'offset']}
+        return MonthBegin(n=1, normalize=self.normalize, **kwds)
+
+    @apply_wraps
+    def apply(self, other):
+        n = self.n
+        dt_in = other
+        # First move to month offset
```
Comment (jreback) [inline pandas/tseries/offsets.py]: same

**Review by jreback — APPROVED** [review]
ok trivial change. this patch is big enough!

**Comment by jbrockmendel** [comment]
> this patch is big enough!

Seriously, yah.  Glad I broke it out from the other one.

**Comment by jreback** [comment]
thanks @jbrockmendel

**Pull request merged**